modules/src/core/ics05_port/context.rs

```diff
@@ -1,70 +1,114 @@
-use crate::core::ics05_port::capabilities::{Capability, CapabilityName, PortCapability};
+use crate::core::ics05_port::capabilities::{Capability, CapabilityName};
 use crate::core::ics05_port::error::Error;
 use crate::core::ics24_host::identifier::PortId;
 use crate::core::ics24_host::path::PortsPath;
 use crate::core::ics26_routing::context::ModuleId;
 use crate::prelude::*;
 
 /// A context supplying all the necessary read-only dependencies for processing any information regarding a port.
-pub trait PortReader: CapabilityReader {
-    /// Return the module_id along with the capability associated with a given port_id
-    fn lookup_module_by_port(&self, port_id: &PortId) -> Result<(ModuleId, PortCapability), Error>;
+pub trait PortCapabilityReader<M: Into<ModuleId>>: CapabilityReader<M> {
+    /// Return the `ModuleId` along with the `Capability` associated with a given `PortId`
+    fn lookup_module_by_port(
+        &self,
+        port_id: PortId,
+    ) -> Result<(ModuleId, Self::Capability), Error> {
+        self.lookup_module(&port_capability_name(port_id))
+            .map(|(module_id, capability)| (module_id, capability))
+    }
 
-    /// Check if the specified port_id is already bounded
+    /// Check if the specified `PortId` is already bound
     fn is_bound(&self, port_id: PortId) -> bool {
-        self.get_capability(&Self::port_capability_name(port_id))
-            .is_ok()
+        self.get_port_capability(port_id).is_ok()
     }
 
-    /// Authenticate a capability key against a port_id by checking if the capability was previously
-    /// generated and bound to the specified port
-    fn authenticate(&self, port_id: PortId, capability: &PortCapability) -> bool {
-        self.authenticate_capability(&Self::port_capability_name(port_id), capability)
-            .is_ok()
+    /// Get the `Capability` associated with the specified `PortId`
+    fn get_port_capability(&self, port_id: PortId) -> Result<Self::Capability, Error> {
+        self.get_capability(&port_capability_name(port_id))
     }
 
-    fn port_capability_name(port_id: PortId) -> CapabilityName {
-        PortsPath(port_id)
-            .to_string()
-            .parse()
-            .expect("PortsPath cannot be empty string")
+    /// Authenticate a `Capability` against the specified `PortId` by checking if the capability
+    /// was previously generated and bound to the specified port
+    fn authenticate_port_capability(
+        &self,
+        port_id: PortId,
+        capability: &Self::Capability,
+    ) -> Result<(), Error> {
+        self.authenticate_capability(&port_capability_name(port_id), capability)
     }
 }
 
-pub trait PortKeeper: CapabilityKeeper + PortReader {
+pub trait PortCapabilityKeeper<M: Into<ModuleId>>:
```

Comment:
`PortCapabilityKeeper` and `ChannelCapabilityKeeper` have been extracted out of `Port/ChannelKeeper`. These traits only have provided methods are provided for convenience.